linkerd/proxy/transport/src/tls/accept.rs

```diff
@@ -163,71 +176,74 @@ where
     }
 }
 
-pub async fn detect(
-    tls_config: Arc<Config>,
-    local_id: identity::Name,
-    mut tcp: TcpStream,
-) -> io::Result<(PeerIdentity, Io)> {
-    const NO_TLS_META: PeerIdentity = Conditional::None(ReasonForNoPeerName::NoTlsFromRemote);
-
-    // First, try to use MSG_PEEK to read the SNI from the TLS ClientHello.
-    // Because peeked data does not need to be retained, we use a static
-    // buffer to prevent needless heap allocation.
-    //
-    // Anecdotally, the ClientHello sent by Linkerd proxies is <300B. So a
-    // ~500B byte buffer is more than enough.
-    let mut buf = [0u8; PEEK_CAPACITY];
-    let sz = tcp.peek(&mut buf).await?;
-    debug!(sz, "Peeked bytes from TCP stream");
-    match conditional_accept::match_client_hello(&buf, &local_id) {
-        conditional_accept::Match::Matched => {
-            trace!("Identified matching SNI via peek");
-            // Terminate the TLS stream.
-            let (peer_id, tls) = handshake(tls_config, PrefixedIo::from(tcp)).await?;
-            return Ok((peer_id, EitherIo::Right(tls)));
-        }
-
-        conditional_accept::Match::NotMatched => {
-            trace!("Not a matching TLS ClientHello");
-            return Ok((NO_TLS_META, EitherIo::Left(tcp.into())));
-        }
-
-        conditional_accept::Match::Incomplete => {}
-    }
-
-    // Peeking didn't return enough data, so instead we'll allocate more
-    // capacity and try reading data from the socket.
-    debug!("Attempting to buffer TLS ClientHello after incomplete peek");
-    let mut buf = BytesMut::with_capacity(BUFFER_CAPACITY);
-    debug!(buf.capacity = %buf.capacity(), "Reading bytes from TCP stream");
-    while tcp.read_buf(&mut buf).await? != 0 {
-        debug!(buf.len = %buf.len(), "Read bytes from TCP stream");
-        match conditional_accept::match_client_hello(buf.as_ref(), &local_id) {
+#[async_trait::async_trait]
+impl Detectable for TcpStream {
+    async fn detected(
+        mut self,
+        tls_config: Arc<Config>,
+        local_id: identity::Name,
+    ) -> io::Result<(PeerIdentity, Io<Self>)> {
+        const NO_TLS_META: PeerIdentity = Conditional::None(ReasonForNoPeerName::NoTlsFromRemote);
+
+        // First, try to use MSG_PEEK to read the SNI from the TLS ClientHello.
+        // Because peeked data does not need to be retained, we use a static
+        // buffer to prevent needless heap allocation.
+        //
+        // Anecdotally, the ClientHello sent by Linkerd proxies is <300B. So a
+        // ~500B byte buffer is more than enough.
+        let mut buf = [0u8; PEEK_CAPACITY];
+        let sz = self.peek(&mut buf).await?;
```

Comment:
This trait is primarily needed so that we can call `TcpStream::peek` which is a specialized API that is (AFAIK) not exposed by `AsyncRead`.